cast/src/tx.rs

```diff
@@ -0,0 +1,360 @@
+use ethers_core::{
+    abi::Function,
+    types::{
+        transaction::eip2718::TypedTransaction, Chain, Eip1559TransactionRequest, NameOrAddress,
+        TransactionRequest, H160, U256,
+    },
+};
+use ethers_providers::Middleware;
+
+use eyre::{eyre, Result};
+use foundry_utils::{encode_args, get_func, get_func_etherscan};
+use futures::future::join_all;
+
+use crate::strip_0x;
+
+pub struct TxBuilder {
+    to: H160,
+    chain: Chain,
+    tx: TypedTransaction,
+    func: Option<Function>,
+    etherscan_api_key: Option<String>,
+}
+
+///
+/// Transaction builder
+/// ```
+/// async fn foo() -> eyre::Result<()> {
+///   use ethers_core::types::{Chain, U256};
+///   use cast::{TxBuilder};
+///   let provider = ethers_providers::test_provider::MAINNET.provider();
+///   let mut builder = TxBuilder::new(&provider, "a.eth", "b.eth", Chain::Mainnet, false).await?;
+///   builder
+///       .gas(Some(U256::from(1)));
+///   let (tx, _) = builder.build();
+///   Ok(())
+/// }
+/// ```
+impl TxBuilder {
+    ///
+    /// Create a new TxBuilder
+    /// `provider` - provider to use
+    /// `from` - 'from' field. Could be an ENS name
+    /// `to` - `to`. Could be a ENS
+    /// `chain` - chain to construct the tx for
+    /// `legacy` - use type 1 transaction
+    pub async fn new<M: Middleware, F: Into<NameOrAddress>, T: Into<NameOrAddress>>(
```

Comment:
Maybe worth storing the `provider` inside the builder? This would allow you to remove the provider arg from `args` and `set_args` which I personally find much cleaner.

Reply:
I didn't want to do that b/c I'd need to deal with lifetimes (have to keep provider by ref) and that was new to me. but, dug in and made it work